Add Support for PCI Devices

JBlaschke opened the pull request:
Complete the work that was started here: https://github.com/JuliaParallel/Hwloc.jl/pull/66

This PR adds full PCI device support to Hwloc.jl. `Hwloc.topology()` now also prints the same output as `lstopo`.

For example:
```
$ julia -e "using Hwloc; topology()" 
Machine (31.06 GB)
    Package L#0 P#0 (31.06 GB)
        NUMANode (31.06 GB)
        L3 (12.0 MB)
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#1 P#1 
                PU L#2 P#1 
                PU L#3 P#5 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#2 P#2 
                PU L#4 P#2 
                PU L#5 P#6 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#3 P#3 
                PU L#6 P#3 
                PU L#7 P#7 
    HostBridge 
        PCI 00:02.0 (VGA)
            GPU "renderD128"
            GPU "card0"
        PCIBridge 
            PCI 01:00.0 (NVMExp)
                Block(Disk) "nvme0n1"
        PCIBridge 
            PCI 72:00.0 (Network)
                Net "wlp114s0"
        PCIBridge 
            PCI 73:00.0 (Other)
                Block "mmcblk0"
```
whereas the `lstopo` outut is:
```
$ lstopo
Machine (31GB total)
  Package L#0
    NUMANode L#0 (P#0 31GB)
    L3 L#0 (12MB)
      L2 L#0 (1280KB) + L1d L#0 (48KB) + L1i L#0 (32KB) + Core L#0
        PU L#0 (P#0)
        PU L#1 (P#4)
      L2 L#1 (1280KB) + L1d L#1 (48KB) + L1i L#1 (32KB) + Core L#1
        PU L#2 (P#1)
        PU L#3 (P#5)
      L2 L#2 (1280KB) + L1d L#2 (48KB) + L1i L#2 (32KB) + Core L#2
        PU L#4 (P#2)
        PU L#5 (P#6)
      L2 L#3 (1280KB) + L1d L#3 (48KB) + L1i L#3 (32KB) + Core L#3
        PU L#6 (P#3)
        PU L#7 (P#7)
  HostBridge
    PCI 00:02.0 (VGA)
    PCIBridge
      PCI 01:00.0 (NVMExp)
        Block(Disk) "nvme0n1"
    PCIBridge
      PCI 72:00.0 (Network)
        Net "wlp114s0"
    PCIBridge
      Block "mmcblk0"
```

vchuravy committed:
WIP: Add support for PCI devices

Co-authored-by: Johannes Blaschke <johannes.blaschke@gmail.com>

vchuravy committed:
fixup! WIP: Add support for PCI devices

vchuravy committed:
Add PCIDevAttr

JBlaschke committed:
Merge branch 'master' into vc/pci

JBlaschke committed:
wip on printing pci data

JBlaschke committed:
Merge branch 'master' into vc/pci

JBlaschke committed:
wip

JBlaschke committed:
extract more pci data

JBlaschke committed:
read dev class id

JBlaschke committed:
wip on inserting process information

JBlaschke committed:
print PCI devices properly now

JBlaschke committed:
get lstopo-style printing

JBlaschke committed:
add coproc subtype

JBlaschke committed:
update list of minimal device types

JBlaschke committed:
add more minimal devices

carstenbauer commented on src/highlevel_api.jl:
```diff
@@ -306,6 +392,6 @@ The quality of the result might depend on the used terminal and might vary betwe
 """
 function topology_graphical()
-    run(`$(lstopo_no_graphics()) --no-io --no-legend --of txt`)
+    run(`$(lstopo_no_graphics()) --no-legend --of txt`)
```
Maybe add an `io` keyword argument based on which we add `--no-io` or not? Sometimes one doesn't care about IO components and they can very much clutter the visualization.

JBlaschke replied:
Done. I'm using `get_io` which is the same kwarg as with `gettopology` and `topology_init`

carstenbauer commented on src/highlevel_api.jl:
```diff
@@ -71,7 +157,7 @@ end
 Prints the system topology as a tree.
 """
-topology() = print_topology(gettopology())
+topology(htopo=nothing) = print_topology(gettopology(htopo))
```
Similar to my other comment, should we introduce an `io` keyword here to give the user the options to opt-in/opt-out of showing IO devices?

(Also, you've added the `minimal` option to `print_topology`. As users aren't really expected to call this function directly, the option isn't really user-facing. Is this intentional?)

JBlaschke replied:
RE the I/O kwarg: I like the idea, but indead I'm passing the `get_io` kwarg through `gettopology` (`topology_init` already has this). So someone sees IO devices by default:
```julia
julia> topo = gettopology(;reload=true)
Hwloc.Object: Machine

julia> topology(topo)
Machine (31.05 GB)
    Package L#0 P#0 (31.05 GB)
        NUMANode (31.05 GB)
        L3 (12.0 MB)
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#1 P#1 
                PU L#2 P#1 
                PU L#3 P#5 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#2 P#2 
                PU L#4 P#2 
                PU L#5 P#6 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#3 P#3 
                PU L#6 P#3 
                PU L#7 P#7 
    HostBridge 
        PCI 00:02.0 (VGA)
            GPU "renderD128"
            GPU "card0"
        PCIBridge 
            PCI 01:00.0 (NVMExp)
                Block(Disk) "nvme0n1"
        PCIBridge 
            PCI 72:00.0 (Network)
                Net "wlp114s0"
        PCIBridge 
            PCI 73:00.0 (Other)
                Block "mmcblk0"
```
or they can set `get_io=false` like so:
```julia
julia> topo = gettopology(;reload=true, get_io=false)
Hwloc.Object: Machine

julia> topology(topo)
Machine (31.05 GB)
    Package L#0 P#0 (31.05 GB)
        NUMANode (31.05 GB)
        L3 (12.0 MB)
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#1 P#1 
                PU L#2 P#1 
                PU L#3 P#5 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#2 P#2 
                PU L#4 P#2 
                PU L#5 P#6 
            L2 (1.25 MB) + L1 (48.0 kB) + Core L#3 P#3 
                PU L#6 P#3 
                PU L#7 P#7 
```

How does that sound?

The minimal classes is a bit of a hack -- it's mainly used the squash the absolutely obscene number of I/O devices on semi-modern apple computers, and clusters. I can remove it if it causes confusion. So `minimal` is only used to filter out the noise and make `print_topology` more like `lstopo`. I've documentat this behaviour in the docstring now:
```julia
julia> print_topology()

Machine (31.05 GB)
    Package L#0 P#0 (31.05 GB)
        NUMANode (31.05 GB)
        L3 (12.0 MB)
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#1 P#1 
                PU L#2 P#1 
                PU L#3 P#5 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#2 P#2 
                PU L#4 P#2 
                PU L#5 P#6 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#3 P#3 
                PU L#6 P#3 
                PU L#7 P#7 
    HostBridge 
        PCI 00:02.0 (VGA)
            GPU "renderD128"
            GPU "card0"
        PCIBridge 
            PCI 01:00.0 (NVMExp)
                Block(Disk) "nvme0n1"
        PCIBridge 
            PCI 72:00.0 (Network)
                Net "wlp114s0"
        PCIBridge 
            PCI 73:00.0 (Other)
                Block "mmcblk0"

julia> print_topology(;minimal=false)

Machine (31.05 GB)
    Package L#0 P#0 (31.05 GB)
        NUMANode (31.05 GB)
        L3 (12.0 MB)
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#1 P#1 
                PU L#2 P#1 
                PU L#3 P#5 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#2 P#2 
                PU L#4 P#2 
                PU L#5 P#6 
            L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#3 P#3 
                PU L#6 P#3 
                PU L#7 P#7 
    HostBridge 
        PCI 00:00.0 (HostBridge)
        PCI 00:02.0 (VGA)
            GPU "renderD128"
            GPU "card0"
        PCI 00:04.0 (SignalProcessing)
        PCIBridge 
            PCI 01:00.0 (NVMExp)
                Block(Disk) "nvme0n1"
        PCI 00:0a.0 (SignalProcessing)
        PCI 00:0d.0 (USB)
        PCI 00:0d.2 (USB)
        PCI 00:0d.3 (USB)
        PCI 00:12.0 (Serial)
        PCI 00:14.0 (USB)
        PCI 00:14.2 (RAM)
        PCI 00:15.0 (SerialBus)
            DMA "dma0chan0"
            DMA "dma0chan1"
        PCI 00:15.1 (SerialBus)
            DMA "dma1chan0"
            DMA "dma1chan1"
        PCI 00:16.0 (Communication)
        PCI 00:19.0 (SerialBus)
            DMA "dma2chan0"
            DMA "dma2chan1"
        PCI 00:19.1 (SerialBus)
            DMA "dma3chan0"
            DMA "dma3chan1"
        PCIBridge 
            PCI 72:00.0 (Network)
                Net "wlp114s0"
        PCIBridge 
            PCI 73:00.0 (Other)
                Block "mmcblk0"
        PCI 00:1f.0 (ISABridge)
        PCI 00:1f.3 (MultimediaAudio)
        PCI 00:1f.4 (SMBus)
        PCI 00:1f.5 (SerialBus)

carstenbauer replied:
Ok, thanks. I'm fine with `minimal` and `get_io=false` (although just `io=false` might be nicer, but whatever I guess).

Just one more thing: I note that the regular (i.e. non-io) printing has changed. Specifically, there seem to be new line breaks for the lines starting with `+`. That is, we now get

```
L2 (1.25 MB)
             + L1 (48.0 kB)
             + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#4
```
instead of the previous, more compact
```
L2 (256.0 kB) + L1 (32.0 kB) + Core L#0 P#0 
                PU L#0 P#0 
                PU L#1 P#1
```
I prefer the old behavior.

JBlaschke replied:
oops! fixed it. Also changed all references of `get_io` to just `io`

carstenbauer commented on src/libhwloc_extensions.jl:
```diff
@@ -0,0 +1,17 @@
+module LibHwlocExtensions
+
+using ..LibHwloc: libhwloc, Hwloc_jll
```
You don't need `Hwloc_jll` here I think?

JBlaschke replied:
Correct! Removed!

carstenbauer commented on src/libhwloc_extensions.jl:
```diff
@@ -0,0 +1,17 @@
+
+using CEnum
+
+function hwloc_pci_class_string(class_id)
```
Why is this in a new file/module and not just in `libhwloc.jl`?

JBlaschke replied:
`libhwloc.jl` is machine generated (modulo a few edge cases that Clang.jl couldn't handle) by `generator.jl`. The intention is that if Hwloc ever chagned its ABI then `libhwloc.jl` would be regenerated, deleting any manual definitions. This definition is not part of Hwloc's `include/hwloc/*` header files, instead it's in `include/private/misc.h` (https://github.com/open-mpi/hwloc/blob/master/include/private/misc.h don't ask me why ...)

Anyway, this is to say that I want to preserve this manual addition to the ABI wrappers in the event of regenerating `libhwloc.jl`

carstenbauer reviewed (commented):
LGTM but see my comments. Maybe you also want to add the PCI device information to the `getinfo` function. `hwloc-info` shows this information as well.

JBlaschke committed:
update Pkg compat

JBlaschke committed:
add more control whether or not to print I/O devices

JBlaschke committed:
Update docs

JBlaschke committed:
fix how settings are filtered through

JBlaschke committed:
remove unnecessary import

JBlaschke commented:
@carstenbauer what do you think?

@vchuravy the CI against nightly Julia is having trouble with converting a `Vector{hwloc_obj_t}` to `Ptr{hwloc_obj_t}` here: https://github.com/JuliaParallel/Hwloc.jl/blob/1d8e48c0b1e5c8d88aa8dfe733a753985d496b3b/src/lowlevel_api.jl#L281

```julia
  conversion to pointer not defined for Vector{Ptr{Hwloc.LibHwloc.hwloc_obj}}
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:35
    [2] unsafe_convert(::Type{Ptr{Ptr{Hwloc.LibHwloc.hwloc_obj}}}, a::Vector{Ptr{Hwloc.LibHwloc.hwloc_obj}})
```

Is this something we should worry about? I can look for a solution, but vote for merging in the meantime.

JBlaschke committed:
try unsafe_wrap instead of convert->copyto

JBlaschke commented:
NVM -- false alarm -- I realized that that code was a bit of a workaround anyway. So switched to using `unsafe_wrap` instead.

@carstenbauer hit "resolve" on all your comments above if I addressed them to your satisfaction.

JBlaschke committed:
update readme

JBlaschke commented:
I've also just refreshed the README

carstenbauer commented:
Looks mostly good to me. Apart from my comment on printing above, you doesn't seem to have addressed the comment
> Maybe you also want to add the PCI device information to the `getinfo`/`topology_info` function. `hwloc-info` shows this information as well.

Want to add this?

I guess it can also slightly misleading that the screenshot for `topology_graphical` is from a different system than the rest of the README. If you can update this (in color) that'd be great. Otherwise, e.g. if on your system you only get black and white output, we can leave it as is.

JBlaschke committed:
oops! fixed it

JBlaschke committed:
get_io -> io

JBlaschke committed:
Add iterator over I/O devices

JBlaschke committed:
Add all devices to getinfo

JBlaschke committed:
Add PCI devices to topology_info

JBlaschke committed:
Update topology_graphical()

JBlaschke commented:
> Looks mostly good to me. Apart from my comment on printing above, you doesn't seem to have addressed the comment
> 
> > Maybe you also want to add the PCI device information to the `getinfo`/`topology_info` function. `hwloc-info` shows this information as well.
> 
> Want to add this?

Oh oooops! I didn't pay enough attention: It looks like I forgot to update the iterators. Fixed it, and now `getinfo` and `topology_info` have PCI device counts also.

> I guess it can also slightly misleading that the screenshot for `topology_graphical` is from a different system than the rest of the README. If you can update this (in color) that'd be great. Otherwise, e.g. if on your system you only get black and white output, we can leave it as is.

Yea, I had left the screenshot in because only my Mac shows colorful output (but no PCI devices?! -- Mac is weird). I have updated the readme to show a non-colorful output, but with PCI.

JBlaschke committed:
Update README to reflect changes to print_topology

carstenbauer commented:
LGTM. I will merge this later today (~ in 6 hours) if there are no objections.
